### DISCUSSION TITLE
AVA hangs after all tests pass

### DISCUSSION THREAD

**Discussion by rrthomas** [avajs/ava#3259] [Q&A] [answered]
I have a small project with a simple test suite, with no async code (that I'm aware of, at least!): https://github.com/ursalang/ark

When I run the tests, they all pass, but AVA hangs and runs indefinitely at 100% CPU.

The output looks like this:

```
  ✔ basics › Concrete values
  ✔ basics › Intrinsics
  ✔ basics › Sequences
  ✔ basics › Conditionals
  ✔ basics › Bare break
  ✔ basics › loop and break
  ✔ basics › let
  ✔ basics › Objects
  ✔ basics › Lists
  ✔ basics › Maps
  ✔ examples › Increment a variable in a loop
  ✔ examples › Sum ints from 1 to 10
  ✔ examples › Call first-class function
  ✔ examples › Factorial (recurse with symbol reference)
  ✔ examples › Factorial (recurse with function argument)
  ✔ examples › Sum list (break result)
  ✔ examples › Sum list (return result)
  ✔ examples › Double list
  ✔ examples › Assign to table
  ✔ examples › Repeated closure
  
  ✘ Timed out while running tests
```

At the end, AVA is still running.

AVA configuration (from `package.json`):

```json
"ava": {
    "typescript": {
      "rewritePaths": {
        "src/": "lib/"
      },
      "compile": "tsc"
    }
  },
  ```

AVA invocation: `npx ava`.
AVA version: 5.3.1

Apologies for not providing a minimal example yet; I will endeavour to whittle it down. I note already that even running one of the tests using e.g. 

```
npx ava --match "Lists" src/basics.test.ts
```

will cause the hang sometimes. The more tests I run, the more often it hangs, it seems. Running one of the two test files:

```
npx ava src/basics.test.ts
```

will often but not always hang.

The code being run is entirely deterministic and non-async, so I can't see any reason why I should get different results on different runs.

I tried running the test suite on two different machines "just in case" and got the same symptoms.

 Thanks for AVA!

**Comment by sindresorhus** [comment]
The problem seems to be with worker threads. Works fine if you use the `--no-worker-threads` flag.

**Comment by rrthomas** [comment]
Thanks ever so much for the quick response and workaround. Is there anything else you need me to do here? It sounds as though you're saying it looks like a problem with AVA?

**Reply by novemberborn** [reply]
@rrthomas thanks for sharing a repository where this is reproducible!

It's slightly better with the `main` branch but I could still get it to hang. Best I can tell when adding log statements is that something goes haywire when explicitly exiting the worker thread. So let's not do that, see https://github.com/avajs/ava/pull/3260.

Could you give that branch a try on your machine, with worker threads re-enabled?

**Reply by rrthomas** [reply]
Sorry, I didn't see this until now. I will try it. Recently I have noticed that I have the problem very occasionally even without worker threads.

**Reply by rrthomas** [reply]
The #3260 branch with worker threads on seems to work fine, thanks!

**Reply by novemberborn** [reply, marked as answer]
This has shipped in AVA 6.

**Reply by rrthomas** [reply]
Many thanks, I'll upgrade.

**Reply by rrthomas** [reply]
I'm still getting occasional hangs with AVA 6. I'll test again whether switching off worker threads fixes it.

**Reply by rrthomas** [reply]
Switching off worker threads again seems to have fixed the problem, although it was rare, so I can't be 100% sure.

**Reply by rrthomas** [reply]
In fact, I'm still getting occasional hangs with AVA 6 with worker threads off.

**Reply by novemberborn** [reply]
I don't know why that would be — we're not doing anything out of the ordinary I think.

**Reply by rrthomas** [reply]
Just sometimes, I get timeouts when my machine is heavily loaded, which is correct: I have kept the default timeout of 10s, and some tests can run longer than that when the machine's busy.

However, I still occasionally get the problem where all the tests are marked as successfully completed, but I still get a timeout error at the end.

**Comment by Kikobeats** [comment]
I noted this is happening to me after v6 is released because my code is using puppeteer and it's spawning a process that is still running after all tests are pasing green.

The ideal solution for this is to control the process lifecycle and kill it before ava exits. I did that and it worked.

However, there are other scenarios where do that is not easy. I was wondering if ava make this easier just waiting to a specific process signal (like SIGTERM) or provide a mecanihsm to do it, something a global teardown:

```js
if (NODE_ENV === 'test') require('ava').teardown(browser.close)
```

or maybe as cli flag? thinking into `ava --force-exit` just to make the old behavior conditional

**Comment by abcfy2** [comment]
I have the same error here. Please help.

**Comment by XenoWad01** [comment]
I also have this issue, for me it occurs 100% of times but I do have async stuff happening.  I am trying to test some trpc endpoints so my environment is Node with esm. Not sure if this is relevant but here is my tsconfig.json seeing as I use typescript with tsx: 
```json
{
  // "include": ["./", "./**/*.ts"],
  "allowSyntheticDefaultImports": true,      
  "compilerOptions": {
    /* Visit https://aka.ms/tsconfig to read more about this file */

    "target": "ES2022",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
   
    "module": "ES2022",
    "moduleResolution": "Bundler",
                          /* Specify a file that bundles all outputs into one JavaScript file. If 'declaration' is true, also designates a file that bundles all .d.ts output. */
    "outDir": "./dist",                                   /* Specify an output folder for all emitted files. */
    "rootDir": ".",
    "baseUrl": ".",
     "allowSyntheticDefaultImports": true,             /* Allow 'import x from y' when a module doesn't have a default export. */
    "esModuleInterop": true,                             /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables 'allowSyntheticDefaultImports' for type compatibility. */
    "forceConsistentCasingInFileNames": true,            /* Ensure that casing is correct in imports. */

    /* Type Checking */
    "strict": true,                                      /* Enable all strict type-checking options. */
    
    "skipLibCheck": true                                 /* Skip type checking all .d.ts files. */
  }
}

```
 I have logged before and after everything and made sure all the tests finish including teardowns + the final after. I am testing with only 1 file which looks like this for reference: 
```ts
import anyTest from 'ava';
import { TypedTestFn, hardcodedReusableTestUsers } from './utils';
import { createTestContext } from './utils';
import { db } from 'db/drizzle';
import { user } from 'db/schema';
import { eq } from 'drizzle-orm';

const test = anyTest as TypedTestFn

test.before(async t => {
    t.context = await createTestContext()
})
test.serial('Create', async t => {
    t.timeout(2000)
    const newUser = await t.context.callers.admin.client.create({
        userData: {
            password: 'ABCD',
            phone: '666424324321231',
            name: 'TESTIMUS MAXIMUS',
            email: 'whatever@sdfafdsafewafsda.com'
        },
        note: 'idk this is optional',
        discountData: {
            flatValue: 10,
            percentage: 1,
        }
    })

    console.log('newUser: ', newUser)
    // t.notThrowsAsync(t.context.callers.admin.client.create({
    //     userData: {
    //         password: 'ABCD',
    //         phone: '666424324321231',
    //         name: 'TESTIMUS MAXIMUS',
    //         email: 'whatever@sdfafdsafewafsda.com'
    //     },
    //     note: 'idk this is optional',
    //     discountData: {
    //         flatValue: 10,
    //         percentage: 1,
    //     }
    // }))
    // t.is(1,1)
    t.pass('All good I hope')
    t.teardown(async () => {
        console.log('doin teardown')
        await t.context.db.delete(user).where(eq(user.id, newUser.userId))
        console.log('teardown done')
    })
} );

test.after.always(async () => {
    console.log('AFTER start')
    await db.delete(user).where(eq(user.phone, hardcodedReusableTestUsers.user.phone))
    await db.delete(user).where(eq(user.phone, hardcodedReusableTestUsers.admin.phone))
    
    console.log('AFTER DONE')
    
})

```
Output: 
```
newUser:  {
  id: 'cnt_01hm0fwg236fstm452gc2mt4ys',
  note: 'idk this is optional',
  discountId: null,
  userId: 'usr_01hm0fwg1mvbtyw3ewz807mzpz',
  createdAt: 2024-01-13T04:04:39.107Z,
  updatedAt: 2024-01-13T04:04:39.107Z,
  user: {
    id: 'usr_01hm0fwg1mvbtyw3ewz807mzpz',
    email: 'whatever@sdfafdsafewafsda.com',
    phone: '666424324321231',
    password: 'redacted',
    name: 'TESTIMUS MAXIMUS',
    isAdmin: false,
    clientId: null,
    createdAt: 2024-01-13T04:04:39.091Z,
    updatedAt: 2024-01-13T04:04:39.091Z
  }
}
doin teardown
teardown done
  ✔ Create (432ms)
AFTER start
AFTER DONE
  
  ✘ Timed out while running tests

  Failed to exit when running tests/runTests.test.ts

  ─

  1 test passed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

This is a small utils file that might be relevant seeing as this is where I declare the function that instantiates my context and declare my types:
```ts
import { type TestFn } from 'ava';
import { db } from '../db/drizzle';
import { client, user } from 'db/schema';
import { appRouter, t, emptyContextMethods } from '../trpc/index';
import { eq } from 'drizzle-orm';


export const createApiCaller = t.createCallerFactory(appRouter)


export const hardcodedReusableTestUsers = {
    user: {
        name: 'UserName1321',
        phone: '1111',
        email: 'user.kekimus1111@test.com',
        password: 'whatever'
    },
    admin: {
        name: 'AdminUserName1230',
        phone: '0000',
        email: 'adminUser.kekimus0000@test.com',
        password: 'test_admin_user_password',
        isAdmin: true
    }
}


export const createTestContext = async (shouldThrowAway: boolean = false) => {

    const [testUserFromDb] = await db.insert(user).values({
        ...hardcodedReusableTestUsers.user
    }).returning()
    const [testAdminUserFromDb] = await db.insert(user).values({
        ...hardcodedReusableTestUsers.admin
    }).returning()

    const [testUserClient] = await db.insert(client).values({
        userId: testUserFromDb.id
    }).returning()
    const [testAdminUserClient] = await db.insert(client).values({
        userId: testAdminUserFromDb.id
    }).returning()

    if(shouldThrowAway) {
        await db.delete(user).where(eq(user.id, testUserFromDb.id))
        await db.delete(user).where(eq(user.id, testAdminUserFromDb.id))
        await db.delete(client).where(eq(client.id, testUserClient.id))
        await db.delete(client).where(eq(client.id, testAdminUserClient.id))
    }

    const testUser = {
        ...testUserFromDb,
        client: testUserClient
    }

    const testAdminUser = {
        ...testAdminUserFromDb,
        client: testAdminUserClient
    }
    return {
        db: db,
        callers: {
            authed: createApiCaller({
                ...emptyContextMethods,
                userAgent: 'TestUserAgent',
                user: testUser,
                db,
            }),
            notAuthed: createApiCaller({
                ...emptyContextMethods,
                userAgent: undefined,
                user: undefined,
                db,
            }),
            admin: createApiCaller({
                ...emptyContextMethods,
                userAgent: 'TestAdminUserAgent',
                user: testAdminUser,
                db
            })

        }
    }

}

const exampleContextSoICanInferTyping = await createTestContext(true)

export type TestContextType = typeof exampleContextSoICanInferTyping

export type TypedTestFn = TestFn<TestContextType>

```

I have the following relevant stuff in package.json:
```json
{
  "name": "@bookd/bookd-api",
  "version": "0.0.1",
  "description": "book'd API Library",
  "main": "app.ts",
  "type": "module",
  "ava": {
    "extensions": {
      "ts": "module",
      "js": true
    },
    "nodeArguments": [
      "--loader=tsx"
    ]
  },
  "scripts": {
    "build": "tsup ./app.ts --format esm --dts",
    "dev": "tsx watch ./app.ts",
    "start": "tsc && npm run dev",
    "test": "NODE_OPTIONS='--loader=tsx --no-warnings' ava ./tests/runTests.test.ts",
    "lint": "eslint . --ext ts --report-unused-disable-directives --max-warnings 0",
    "studio": "drizzle-kit studio --port 5666 --verbose --config ./db/drizzle.config.ts ",
    "migration:push": "tsx ./db/migrate.ts",
    "migration:generate": "drizzle-kit generate:pg --schema ./db/schema --out ./migrations"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@trpc/client": "^10.38.3",
    "@trpc/server": "^10.38.3",
    "argon2": "^0.31.2",
    "colors": "^1.4.0",
    "cookie": "^0.6.0",
    "cors": "^2.8.5",
    "drizzle-orm": "^0.29.1",
    "drizzle-zod": "^0.5.1",
    "jsonwebtoken": "^9.0.2",
    "ms": "^2.1.3",
    "pg": "^8.11.3",
    "postgres": "3.3.5",
    "superjson": "^2.2.1",
    "ulid": "2.3.0",
    "zod": "^3.22.2"
  },
  "devDependencies": {
    "@types/cookie": "^0.6.0",
    "@types/cors": "^2.8.17",
    "@types/express": "^4.17.17",
    "@types/jsonwebtoken": "^9.0.5",
    "@types/ms": "^0.7.34",
    "@types/node": "^20.6.0",
    "@types/pg": "^8.10.9",
    "@typescript-eslint/eslint-plugin": "^6.7.3",
    "@typescript-eslint/parser": "^6.7.3",
    "ava": "^6.0.1",
    "dotenv": "16.3.1",
    "drizzle-kit": "^0.20.6",
    "eslint": "^8.50.0",
    "eslint-config-custom": "*",
    "ts-node": "^10.9.2",
    "tsconfig": "*",
    "tsup": "^8.0.1",
    "tsx": "^4.6.2",
    "typescript": "^5.2.2"
  }
}

```
node version:
```
v20.11.0
```

Hope this helps. Let me know if you need more info from me. I can reproduce this 100% of times. 
Also shouldnt this be an issue instead of a discussion?

**Reply by XenoWad01** [reply]
I have tried with or without `.serial` and/or `.always` those made no difference at all.

**Comment by sangaline** [comment]
I'm also getting this with AVA v6.0.1 and worker threads disabled. I'm using Puppeteer as well, but am closing the browser in an async `after.always` hook and the chromium processes are no longer running when AVA is stuck after the tests. I would say it happens roughly 25% of the time for me.

Does anybody have a workaround?

**Comment by kalda341** [comment]
I'm experiencing this too - has anyone figured out a way to solve this?

**Comment by danpalmer** [comment]
I get this on every run, `--no-worker-threads` doesn't seem to make any difference. Node 21.5, ava 6.1.1.

**Comment by karanikiotis** [comment]
Any updates on this? It happens to me too...

**Comment by novemberborn** [comment]
In response to the original post in this thread, AVA 6 removed the previous behavior of forcibly exiting the worker thread. This may cause problems with test environments that have open handles after tests complete. See https://github.com/avajs/ava/blob/main/docs/08-common-pitfalls.md#timeouts-because-a-file-failed-to-exit

**Comment by AkbarHabeeb** [comment]
@novemberborn I'm trying to migrate from v2 of AVA to v6.0.0
Mycode has worker threads to do the logging, on updating the version to 6. Its not running properly.

1. If workerThreads flag is set as false in AVA config, then verbose says all tests passed but then process hangs abruptly for few seconds and then connection drops with message _Process Exiting Due to  0 null_
2. If workerThreads flag set as true, then test is not even running. We would definitely looking to taking advantage of AVA's worker threads

Can you please share your views on this? it will help us a lot. Thanks

**Comment by victorteokw** [comment]
Unable to import this
```
/**
 * Register a function to be called when AVA has completed a test run without uncaught exceptions or unhandled rejections.
 *
 * Completion handlers are invoked in order of registration. Results are not awaited.
 */
declare const registerCompletionHandler: (handler: () => void) => void;
export {registerCompletionHandler};
```
<img width="370" alt="Screenshot 2024-11-03 at 20 39 54" src="https://github.com/user-attachments/assets/e4b9f71f-97ea-429f-a932-26dd0c44a101">
